### PR TITLE
Hip on amd hardware

### PR DESCRIPTION
**Pull request by Rombur**
Here is the changes I needed to get Kokkos to compile.

### CODE REVIEW

File: core/src/HIP/Kokkos_HIP_Atomic.hpp [hunk 1]
```diff
@@ -287,210 +284,285 @@ namespace Kokkos {
...
+}
 
-    return oldval.t ;
-  }
+KOKKOS_INLINE_FUNCTION
+int64_t atomic_fetch_sub(int64_t *dest, int64_t const &val)
+{
+  return static_cast<int64_t>(
+      atomicAdd(reinterpret_cast<unsigned long long *>(dest),
+                -reinterpret_cast<unsigned long long const &>(val)));
```
Comment (dalg24) [inline core/src/HIP/Kokkos_HIP_Atomic.hpp:396]: Are you worried about inlining? Why not `return atomic_fetch_add*dest, -sub);`

File: core/src/HIP/Kokkos_HIP_Atomic.hpp [hunk 2]
```diff
@@ -287,210 +284,285 @@ namespace Kokkos {
...
+
+  do
+  {
+    assume = oldval;
+    newval = assume & 0x7fffff00 + ((assume & 0xff) - val) & 0xff;
+    oldval = hc::atomic_compare_exchange_unsigned((unsigned int *)dest, assume,
+                                                  newval);
+  } while (assume != oldval);
+
+  return oldval;
```
Comment (dalg24) [inline core/src/HIP/Kokkos_HIP_Atomic.hpp:414]: Where does that come from?

File: core/src/HIP/Kokkos_HIP_KernelLaunch.hpp [hunk 3]
```diff
@@ -141,9 +141,11 @@ struct HIPParallelLaunch<
...
              block.y, block.z, shmem);
       printf("Pre Launch Error: %s\n", hipGetErrorName(hipGetLastError()));
-      hip_parallel_launch_local_memory<DriverType, MaxThreadsPerBlock,
-                                       MinBlocksPerSM>
-          <<<grid, block, shmem, hip_instance->m_stream>>>(driver);
+
+      hipLaunchKernelGGL(
+          hip_parallel_launch_local_memory<DriverType, MaxThreadsPerBlock,
+                                           MinBlocksPerSM>,
+          shmem, hip_instance->m_stream, driver);
```
Comment (dalg24) [inline core/src/HIP/Kokkos_HIP_KernelLaunch.hpp]: No `grid` and `block` arguments?

File: core/src/HIP/Kokkos_HIP_Atomic.hpp (hunk 1 above)
Comment (Rombur) [reply core/src/HIP/Kokkos_HIP_Atomic.hpp:396]: You can't do that. HIP does not support `atomicAdd` with `int64_t`

File: core/src/HIP/Kokkos_HIP_Atomic.hpp (hunk 2 above)
Comment (Rombur) [reply core/src/HIP/Kokkos_HIP_Atomic.hpp:414]: Here https://github.com/ORNL-CEES/kokkos/pull/10/files#diff-ed152c9ca6d8623c27b861a713236a76L319 I just uncommented the code

File: core/src/HIP/Kokkos_HIP_KernelLaunch.hpp (hunk 3 above)
Comment (Rombur) [reply core/src/HIP/Kokkos_HIP_KernelLaunch.hpp]: yeah they should be there, will add

**Comment by Rombur** [comment]
So what should we do with this PR?

**Comment by dalg24** [comment]
> So what should we do with this PR?

Merge?

**Comment by Rombur** [comment]
Do you want to understand why I need this and you don't before merging?

**Pull request merged**